sigs/jvm/rfcs/20191023-tensorflow-keras.md

```diff
@@ -0,0 +1,342 @@
+| **Sponsor**   | Karl Lessard (karl.lessard@gmail.com)                 |
+| **Updated**   | 2019-10-23                                          |
+
+## Objective
+
+To create a high-level API for building, training, and experimentation 
+with TensorFlow models in Java, supporting the syntax and
+standards of [tf.keras](https://www.tensorflow.org/api_docs/python/tf/keras).
+
+The design of the API and summary written here is based on
+the open source code and documentation of 
+[tf.keras](https://www.tensorflow.org/api_docs/python/tf/keras).
+
+## Motivation
+
+The Keras deep learning API is now a widely-accepted standard; greatly simplfying
+model implementations, and enabling efficient experimentation as well as rapid
+production of trained models with cross-platform framework support. It has seen
+adoption across the TensorFlow ecosystem (TF 2.0, tfjs).
+
+Currently, the TensorFlow Java client requires direct manipulation of the
+graph operand-by-operand to build and train models. While
+`Graph` objects can be seralized and loaded in Java, the lack of a standard `Model`
+framework makes TensorFlow Java projects custom, and slow to develop.
+
+## User Benefit
+
+Users of TensorFlow-Java focused on experimentation, production, and depolyment
+of deep learning models will benefit greatly from the Keras API's conciseness and portability.
+
+Combining the wide support of the JVM ecosystem and development environment with a 
+practical interface to TensorFlow Java will allow users to easily integrate deep learning
+into their existing development and production workflows.
+
+## Design Proposal
+
+In building this framework, the goal is to closely match the behaviors and workflows
+Keras supports in its Python API, with a well designed Java API that can be easily
+integrated into existing projects.
+
+### Overview
+
+Keras is built around a few core abstractions which comprise much of what is needed to build
+deep learning models. These include layers, models, optimizers, activations, losses, metrics,
+regularizers, and initializers.
+
+To make our code clearer and more manageable, we distinguish explicitly between two steps:
+Graph building (even in eager mode, graph "definition"), and Graph execution.
+
+
+
+### Layers
+Layers represent arbitrary graph computations on tensors. They are composable; meant to be "stacked"
+on each other to build neural networks. Thus they must encode *differentiable* functions (i.e. built
+using TF Operations).
+
+Layers are defined by (1) a set of weights, and (2) a differentiable "transform" function which uses
+these weights to transform input tensor(s). For example, a `Dense` layer contains a kernel
+weight matrix *w_k* and a bias weight matrix *w_b*, and transforms an input *x* using the
+function *x* -> *f*(*w_k* x + *w_b*) for a user-specified activaition *f*.
+
+#### Defining Layers
+To define a new type of layer, users can override 3 methods from the base `Layer<T>` class.
```

Comment:
Just as a proposal here opened to discussion, instead of relying on inheritance for customizing layers, could we use composition and a factory? something like
```java
public interface LayerBackend<T> {

    void build(Ops tf, Shape inputShape);

    Operand<T> call(Ops tf, Operand<T>... operands);
}

final class DenseLayerBackend<T> implements LayerBackend<T> {
   ...
}

final class Layers {

    public static Layer<T> create(Shape shape, DataType<T> dtype, LayerBackend<T> backend) {
        return Layer.build(shape, dtype, backend);
    }

    public static Layer<T> createDense(Shape shape, DataType<T> dtype) {
        return layerOf(shape, dtype, new DenseLayerBackend<T>());
    }
}
```

Reply:
I definitely like the idea of the `Layers` factory methods. I do see where you're coming from in terms of composition: subclassing layers is very much (likely) about that. (Please correct me if I'm missing something about your intent.)

But many Python layer implementations I've seen subclass mainly to share internal Layer state.

The "Model/Layer [Subclassing API](https://www.tensorflow.org/guide/keras/custom_layers_and_models)" is by now widely understood and used. Since that API is standard, and the subclassing semantics are simple, my inclination is to stick with it here. This makes it easier for users who port custom layer implementations from python, with deep class hierarchies. Thoughts?